[WIP] generalize methods further by creating default method

josephsawaya opened the pull request:
This PR refactors the methods interface further with the creation of a DefaultMethod struct that more specific methods embed to simplify common tasks between the methods such as getting and setting the method's target. 

This PR also changes the configuration file to have fetched create the method structs itself rather than to have viper do that, by changing the configuration file to accept a list of the methods with a kind field specifying the type of the method.

Signed-off-by: Joseph Sawaya <jsawaya@redhat.com

Commit:
generalize methods further by creating default method

This commit refactors the methods interface
further with the creation of a DefaultMethod struct
that more specific methods embed to simplify common
tasks between the methods such as getting and
setting the method's target.

This commit also changes the configuration file
to have fetched create the method structs itself
rather than to have viper do that, by changing
the configuration file to accept a list of the
methods with a kind field specifying the type of
the method.

Signed-off-by: Joseph Sawaya <jsawaya@redhat.com>

Commit:
Move runChangesConcurrent to defaultMethod

This commit moves the runChangesConcurrent definition
to the DefaultMethod struct so every method does not
have to define it.

Signed-off-by: Joseph Sawaya <jsawaya@redhat.com>

sallyom commented:
awesome! we'll bring this in b4 merging the refactor, +1000

sallyom commented:
In the PR I changed `ConfigTarget` to `ConfigReload` and `configTarget` to `configReload`, I believe that's the only conflict. 
I brought these changes in but something wasn't working quite right - we'll get it working tomorrow!

Commit:
Create currentToLatest in DefaultMethod

This commit creates the currentToLatest function
in the DefaultMethod to be used in the Process
functions of the specific methods.

Signed-off-by: Joseph Sawaya <jsawaya@redhat.com>

Commit:
fix log message

Signed-off-by: Joseph Sawaya <jsawaya@redhat.com>
